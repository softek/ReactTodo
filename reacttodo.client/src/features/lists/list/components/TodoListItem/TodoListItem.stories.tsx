import type { Meta, StoryObj } from '@storybook/react';
import {
  expect, fn, userEvent, waitFor, within,
} from '@storybook/test';
import { TodoListItem } from './TodoListItem';

const meta = {
  title: 'Features/TodoListItem',
  component: TodoListItem,
  parameters: {
    layout: 'centered',
  },
  argTypes: {
  },
  args: {
    onDelete: fn(),
    onToggle: fn(),
    onRename: fn(),
  },
} satisfies Meta<typeof TodoListItem>;

export default meta;
type Story = StoryObj<typeof meta>;

export const IncompleteTask: Story = {
  args: {
    task: { id: '1', name: 'Task 1', isComplete: false },
  },
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const checkbox = canvas.getByRole('checkbox');
    await expect(checkbox).toBeInTheDocument();
    await userEvent.click(checkbox);
    await waitFor(() => expect(args.onToggle).toHaveBeenCalled());

    const editButton = canvas.getByRole('button', { name: /Edit/i });
    await expect(editButton).toBeInTheDocument();
    await userEvent.click(editButton);

    const inputBox = canvas.getByRole('textbox');
    await expect(inputBox).toBeInTheDocument();
    await expect(inputBox).toHaveValue('Task 1');

    const saveButton = canvas.getByRole('button', { name: /Save/i });
    userEvent.clear(inputBox);
    await userEvent.click(saveButton);
    await expect(inputBox).toBeInTheDocument();

    await userEvent.type(inputBox, 'Renamed Task 1');
    await userEvent.click(saveButton);
    await waitFor(() => expect(args.onRename).toHaveBeenCalledWith('Renamed Task 1'));
    await expect(inputBox).not.toBeInTheDocument();
  },
};

export const CompletedTask: Story = {
  args: {
    task: { id: '1', name: 'Task 1', isComplete: true },
  },
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const deleteButton = canvas.getByRole('button', { name: /Delete/i });
    await expect(deleteButton).toBeInTheDocument();
    await userEvent.click(deleteButton);
    await waitFor(() => expect(args.onDelete).toHaveBeenCalled());

    const checkbox = canvas.getByRole('checkbox');
    await expect(checkbox).toBeInTheDocument();
    await userEvent.click(checkbox);
    await waitFor(() => expect(args.onToggle).toHaveBeenCalled());
  },
};
