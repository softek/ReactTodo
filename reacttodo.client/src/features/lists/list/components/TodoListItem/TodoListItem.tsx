/* eslint-disable react/jsx-props-no-spreading */
import React, { useState } from 'react';
import {
  FormGroup, Input, Label, ListGroupItem,
} from 'reactstrap';
import { Controller, useForm } from 'react-hook-form';
import { Button } from '../../../../../components';
import { TodoTask } from '../../../../../data';

export type TodoListItemProps = {
  task: TodoTask;
  onDelete: () => void;
  onToggle: () => void;
  onRename: (name: string) => void;
}

type RenameItemProps = {
  task: TodoTask;
  onRename: (name: string) => void;
  onCancel: () => void;
}

type RenameInputs = {
  name: string;
}

function RenameItem({
  task, onRename, onCancel,
}: RenameItemProps) {
  const { control, handleSubmit, formState: { errors } } = useForm<RenameInputs>({
    defaultValues: { name: task.name },
  });

  const onSubmit = (data: { name: string }) => {
    onRename(data.name);
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Controller
        name="name"
        control={control}
        rules={{ required: true }}
        render={({ field }) => (
          <Input
            invalid={!!errors.name}
            {...field}
            autoFocus
            type="text"
          />
        )}
      />
      <Button
        label="Save"
        type="primary"
        onClick={handleSubmit(onSubmit)}
      />
      <Button
        label="Cancel"
        type="secondary"
        onClick={onCancel}
      />
    </form>
  );
}

export function TodoListItem({
  task,
  onDelete,
  onToggle,
  onRename,
}: TodoListItemProps) {
  const textClass = task.isCompleted ? 'text-decoration-line-through' : '';
  const [isEditing, setIsEditing] = useState(false);

  const handleRename = (name: string) => {
    onRename(name);
    setIsEditing(false);
  };

  const id = `task-completion-${task.id}`;

  return (
    <ListGroupItem>
      {
        isEditing
          ? (
            <RenameItem task={task} onRename={handleRename} onCancel={() => setIsEditing(false)} />
          )
          : (
            <>
              <FormGroup
                check
                inline
              >
                <Input
                  id={id}
                  type="checkbox"
                  checked={task.isCompleted}
                  onClick={onToggle}
                />
                <Label check for={id} className={textClass}>
                  {task.name}
                </Label>
              </FormGroup>
              {task.isCompleted && (
                <Button
                  label="Delete"
                  type="danger"
                  onClick={onDelete}
                />
              )}
              {!task.isCompleted && (
                <Button
                  label="Edit"
                  type="secondary"
                  onClick={() => setIsEditing(true)}
                />
              )}
            </>
          )
      }
    </ListGroupItem>
  );
}
