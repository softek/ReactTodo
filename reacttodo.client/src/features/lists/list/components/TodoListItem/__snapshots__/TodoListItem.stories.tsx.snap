// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/TodoList/TodoListItem CompletedTask play-test 1`] = `
<li class="list-group-item">
  <div class="form-check form-check-inline">
    <input id="task-completion-1"
           type="checkbox"
           class="form-check-input"
           checked
    >
    <label for="task-completion-1"
           class="text-decoration-line-through form-check-label"
    >
      Task 1
    </label>
  </div>
  <button type="button"
          class="btn btn-danger"
          aria-label="Delete"
  >
    Delete
  </button>
</li>
`;

exports[`Features/TodoList/TodoListItem IncompleteTask play-test 1`] = `
<li class="list-group-item">
  <div class="form-check form-check-inline">
    <input id="task-completion-1"
           type="checkbox"
           class="form-check-input"
    >
    <label for="task-completion-1"
           class="form-check-label"
    >
      Task 1
    </label>
  </div>
  <button type="button"
          class="btn btn-secondary"
          aria-label="Edit"
  >
    Edit
  </button>
</li>
`;
